src/cryptography/hazmat/backends/openssl/x509.py

```diff
@@ -235,7 +235,12 @@ def parse(self, backend, x509_obj):
                     )
             else:
                 d2i = backend._lib.X509V3_EXT_d2i(ext)
-                assert d2i != backend._ffi.NULL
+                if d2i == backend._ffi.NULL:
+                    backend._consume_errors()
+                    raise ValueError(
+                        "The {0} extension appears to be corrupt".format(oid)
```

Comment:
If X509V3_EXT_d2i fails it's because OpenSSL has no idea how to decode the extension, which can only occur with invalid extension payloads or extensions that it doesn't know anything about (e.g. a custom defined one).